File: app/src/main/java/ru/vincetti/vimoney/utils/LogicMath.java
package ru.vincetti.vimoney.utils;

import java.util.List;

import ru.vincetti.vimoney.MyApp;
import ru.vincetti.vimoney.data.AppExecutors;
import ru.vincetti.vimoney.data.models.AccountListModel;
import ru.vincetti.vimoney.data.models.AccountModel;
import ru.vincetti.vimoney.data.sqlite.AppDatabase;

public class LogicMath {
    // set correct account (accID) balance
    public static void accountBalanceUpdateById(int accId) {
        AppDatabase mDb = AppDatabase.getInstance(MyApp.getAppContext());
        accountBalanceUpdateById(mDb, accId);
    }

    // set correct account (accID) balance with DB
    public static void accountBalanceUpdateById(AppDatabase mDb, int accId) {
        AppExecutors.getsInstance().diskIO().execute(() -> {
            float sum = mDb.transactionDao().loadSumByCheckId(accId);
            mDb.accountDao().updateSumByAccId(accId, sum);
        });
    }

    // All accounts update
    public static void accountBalanceUpdateAll() {
        AppDatabase mDb = AppDatabase.getInstance(MyApp.getAppContext());
        accountBalanceUpdateAll(mDb);
    }

    // All accounts update
    public static void accountBalanceUpdateAll(AppDatabase mDb) {
        AppExecutors.getsInstance().diskIO().execute(() -> {
            List<AccountModel> accounts = mDb.accountDao().loadAllAccountsList();
            accountBalanceUpdateAll(mDb, accounts);
        });
    }

    public static void accountBalanceUpdateAll(AppDatabase mDb, List<AccountModel> accounts) {
        AppExecutors.getsInstance().diskIO().execute(() -> {
            for (AccountModel account : accounts) {
                int accountId = account.getId();
                float sum = mDb.transactionDao().loadSumByCheckId(accountId);
                mDb.accountDao().updateSumByAccId(accountId, sum);
            }
        });
    }

    // Math allUser balance
    public static int userBalanceChange(List<AccountListModel> accList) {
        int bal = 0;
        for (AccountListModel o : accList) {
            bal += o.getSum();
        }
        return bal;
    }
}

File: app/src/main/java/ru/vincetti/vimoney/utils/LogicMath.kt
package ru.vincetti.vimoney.utils

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import ru.vincetti.vimoney.data.repository.TransactionRepo
import ru.vincetti.vimoney.data.sqlite.AccountDao
import ru.vincetti.vimoney.data.sqlite.AppDatabase
import ru.vincetti.vimoney.data.sqlite.TransactionDao

/** Set correct account (accID) balance. */
suspend fun accountBalanceUpdateById(
    transactionDao: TransactionDao,
    accountDao: AccountDao,
    accId: Int
) {
    withContext(Dispatchers.IO) {
        val sum = transactionDao.loadSumByCheckId(accId)
        accountDao.updateSumByAccId(accId, sum)
    }
}

/** Set correct all account balance. */
suspend fun accountBalanceUpdateAll(db: AppDatabase) {
    withContext(Dispatchers.IO) {
        val accDao = db.accountDao()
        val accounts = accDao.loadAllAccounts()
        val transRepo = TransactionRepo(db)
        accounts?.forEach {
            val sum = transRepo.loadCheckSum(it.id)
            accDao.updateSumByAccId(it.id, sum)
        }
    }
}

/** Math all user balance. */
suspend fun userBalanceUpdate(accountDao: AccountDao): Int {
    var balance = 0
    val accounts = accountDao.loadAllAccounts()
    accounts?.let { accs ->
        accs.filter { it.needAllBalance && !it.isArchive }.forEach { balance += it.sum }
    }
    return balance
}
